### ISSUE TITLE
Proposal to change allocation model for memory/CPU

### ISSUE THREAD

**Issue by r2k1** [opencost/opencost#1815]
# Opencost model change

This pull request proposes changes to the current OpenCost model due to potential overestimation of pod cost up to 200% in worst cases. The proposal aims to better estimate pod allocation and accurately represent the state of the node.

## Current Allocation Model

Currently, the allocation model is based on memory consumption, and the cost is calculated according to the following formula:

```
pod_memory = max(pod_memory_used, pod_memory_request) * node_memory_cost
``` 

Due to the fact the cost is based on max(used, request) the cost of every pod can be overestimated up to 200% in worst cases.

To showcase it, here is a few examples:

Assuming the cluster consist from 1 node and 2 pods allocated on the node:

### Example 1
```
node 1GB

pod1_used 1GB
pod1_requested 0GB
pod1_allocated = max(1GB, 0GB) = 1GB

pod2_used 0GB
pod2_requested 1GB
pod2_allocated = max(0GB, 1GB) = 1GB

Node capacity: 1GB
Node used: 1GB
Node requested: 1GB

Node available: 0GB
Node available for resource request: 0GB

Estimated Total Workload: (1GB + 1GB) = 2GB
Estimated Idle: 1GB - 2GB = -1GB
```

### Example 2
```
node 1GB

pod1_used 0.5GB
pod1_requested 0GB
pod1_allocated = max(0.5GB, 0GB) = 0.5GB

pod2_used 0GB
pod2_requested 0.5GB
pod2_allocated = max(0GB, 0.5GB) = 0.5GB

Node capacity: 1GB
Node used: 0.5GB
Node requested: 0.5GB

Node available: 0.5GB
Node available for resource request: 0.5GB

Estimated Total Workload: (0.5GB + 0.5GB) = 1GB
Estimated Idle: 1GB - 1GB = 0GB
```

### Example 3
```
node 1GB

pod1_used 0.2GB
pod1_requested 0.4GB
pod1_allocated = max(0.2GB, 0.4GB) = 0.4GB

pod2_used 0.5GB
pod2_requested 0.1GB
pod2_allocated = max(0.5GB, 0.1GB) = 0.5GB

Node capcity: 1GB
Node used: 0.7GB
Node requested: 0.4GB

Node available: 0.3GB
Node available for resource request: 0.6GB

Estimated Total Workload: (0.4GB + 0.5GB) = 0.9GB
Estimated Idle: 1GB - 0.9GB = 0.1GB
```

As you can see from the example total pod cost can exceed the cost of cluster. And node idle cost may result in negative values (which is probably ceiled at 0 in this cases) and doesn't represent the state of the node correctly.

# Proposal

The proposed solution is to use both pod real usage and pod requests to estimate pod allocation. Instead of directly using max(request, used) to determine pod consumption, the proposal suggests splitting allocated node usage (calculated as the maximum between the sum of all pod requests and real resource usage) between pods according to their max(request, used) consumption.

Each node has two constraints for the same resource: physical memory available for pods and the amount of memory available for pod requests. The "requested" constraint is controlled by kube scheduler, which doesn't allow scheduling new pods on the node if the sum of all pod requests is equal to the node's available memory for request.

**By splitting the allocated node usage between pods according to their max(request, used) consumption, the proposal ensures that the total cost of all pods running on the node will match the cost of the node's utilization portion**. This approach will never exceed 100% of the node cost and accurately represent the state of the node.

The idea is that the node has multiple constraints, and no more workload can be run on the node if at least one of the constraints is reached. For example, a node may have 8GB of memory and utilize only 1GB from it, but if the sum of all pods' requests is 8GB, then no more pods with a memory request can be scheduled on the node. So the pods that runs on the node are responsible for the 100% of the node cost.

```
node_requested_memory = sum(pod_request_memory) by (node)
node_allocated_memory = max(node_used_memory, node_requested_memory)
node_idle_memory = node_capacity_memory - node_allocated_memory

pod_score_memory = max(pod_request_memory, pod_used_memory)
pod_score_relative_memory = pod_score_memory / (sum(pod_score_memory) by (node)) // this will result in values between 0 and 1, and sum of all pods on the node will be equal 1
pod_allocated_memory = node_allocated_memory * pod_score_relative
```

After applying this logic for examples before the next result should be achieved:

### Example 1
```
node 1GB

pod1_used 1GB
pod1_requested 0GB
pod1_score = max(1GB, 0GB) = 1GB

pod2_used 0GB
pod2_requested 1GB
pod2_score = max(0GB, 1GB) = 1GB

pod1_score_relative = 1GB/(1GB + 1GB) = 0.5
pod2_score_relative = 1GB/(1GB + 1GB) = 0.5
pod1_allocated = 0.5 * max((0GB+1GB), (1GB+0GB)) = 0.5GB
pod2_allocated = 0.5 * max((0GB+1GB), (1GB+0GB)) = 0.5GB

Node capacity: 1GB
Node used: 1GB
Node requested: 1GB

Node available: 0GB
Node available for resource request: 0GB

Estimated Total Workload: (1GB + 1GB) = 2GB
Estimated Idle: 1GB - 2GB = -1GB
```

### Example 2
```
node 1GB

pod1_used 0.5GB
pod1_requested 0GB
pod1_score = max(0.5GB, 0GB) = 0.5GB

pod2_used 0GB
pod2_requested 0.5GB
pod2_score = max(0GB, 0.5GB) = 0.5GB

pod1_score_relative = 0.5GB/(0.5GB + 0.5GB) = 0.5
pod2_score_relative = 0.5GB/(0.5GB + 0.5GB) = 0.5
pod1_allocated = 0.5 * max((0GB+0.5GB), (0.5GB+0GB)) = 0.25GB
pod2_allocated = 0.5 * max((0GB+0.5GB), (0.5GB+0GB)) = 0.25GB

Node capacity: 1GB
Node used: 0.5GB
Node requested: 0.5GB

Node available: 0.5GB
Node available for resource request: 0.5GB

Estimated Total Workload: (0.25GB + 0.25GB) = 0.5GB
Estimated Idle: 1GB - 0.5GB = 0.5GB
```

### Example 3
```
node 1GB

pod1_used 0.2GB
pod1_requested 0.4GB
pod1_score = max(0.2GB, 0.4GB) = 0.4GB

pod2_used 0.5GB
pod2_requested 0.1GB
pod2_score = max(0.5GB, 0.1GB) = 0.5GB

pod1_score_relative = 0.4GB/(0.4GB + 0.5GB) = 0.444
pod2_score_relative = 0.5GB/(0.4GB + 0.5GB) = 0.555
pod1_allocated = 0.444 * max((0.2GB+0.5GB), (0.4GB+1GB)) = 0.444*0.7GB = 0.31GB
pod2_allocated = 0.555 * max((0.2GB+0.5GB), (0.4GB+1GB)) = 0.555*0.7GB = 0.39GB

Node capacity: 1GB
Node used: 0.7GB
Node requested: 0.4GB

Node available: 0.3GB
Node available for resource request: 0.6GB

Estimated Total Workload: (0.31GB + 0.39GB) = 0.7GB
Estimated Idle: 1GB - 0.7GB = 0.3GB
```

I can go more into details with some examples of prometheus queries how to achieve the result

**Comment by mattray** [comment]
@r2k1 brought this up in today's OpenCost working group meeting and I suggested he write it up for discussion and explanation. I hoped someone more familiar with the low-level calculations could provide feedback.

**Comment by michaelmdresser** [comment]
This is a fascinating proposal. I certainly understand its technical merit (cc @nikovacevic on "totals" coefficients as a similar concept). It's a pretty big shift, though, so I think we should loop in at least @dwbrown2 for commentary.

**Comment by michaelmdresser** [comment]
And on the more technical side, "proportional asset resource costs" may be relevant here: https://github.com/opencost/opencost/pull/1819

**Comment by michaelmdresser** [comment]
I'll actually back up a step, because I just realized something. Kubecost accounts for this discrepancy between "billed asset costs" and "allocated resource costs" using idle. For most users of K8s, they have > 0 idle because they tend to overrequest resources both for workloads and clusters. But once things get tightened down, or requests are intentionally set lower, Kubecost's idle can (validly) go negative to make sure that "allocated" max(request, usage) costs are in line with node costs.

I know we're currently moving some of these concepts into OpenCost, so we may have a much better story here soon without changing the fundamental "allocation" model.

**Comment by lordofire** [comment]
This is a pretty cool discussion. Just 2 cents from me:

Firstly, from a practical k8s cluster management perspective, I think there is a general practices of setting mem req=limit to avoid potential OOM killing issues node-wise when we have memory over-provisioned, like mentioned [here](https://home.robusta.dev/blog/kubernetes-memory-limit). Therefore, I would not have too much concern on this issue for memory in general.

Secondly, from the calculation perspective, I do not 100% agree on the proposal above. I think the requested resource cost of a pod should always be billing by that pod itself, even though it keeps idling without being able to occupy the resource. In the case of usage, I think it is a separate dimension compared to the request, like the following:

```
Node A
Pod a, b
total_node_cost = pod_a_requested_cost + pod_b_requested_cost + node_request_idle_cost
total_node_cost = pod_a_used_cost + pod_b_used_cost + node_used_idle_cost
```

To combine both dimensions together, I felt it would be great to have the following formula. The general idea is that, all the existing request resources directly curve out all the cost of nodes, and remaining cost will be shared either directly or proportionally based on the over-used percentage of the node resources. 

```
pod_a_delta_cost = max(pod_a_used_cost, pod_a_requested_cost) - pod_a_requested_cost
pod_b_delta_cost = max(pod_b_used_cost, pod_b_requested_cost) - pod_b_requested_cost

if pod_a_delta_cost + pod_b_delta_cost > node_request_idle_cost:
  pod_a_cost = pod_a_requested_cost + node_request_idle_cost * pod_a_delta_cost / (pod_a_delta_cost + pod_b_delta_cost)
  pod_b_cost = pod_b_requested_cost + node_request_idle_cost * pod_b_delta_cost / (pod_a_delta_cost + pod_b_delta_cost)
else:
  pod_a_cost = pod_a_requested_cost + pod_a_delta_cost
  pod_b_cost = pod_b_requested_cost + pod_b_delta_cost
```

Happy to spark more discussions on this topic, and feel free to let me know any comments.

**Comment by r2k1** [comment]
@lordofire, Setting req=limit can be a commonly followed practice, but it may not be realistic to expect all users to adhere to it.

Regarding the importance of pod request versus pod real usage. In my view, both dimensions are equally significant for cost allocation purposes, and combining them poses the main challenge.

Allow me to provide a counter-example to support my point. Suppose we have pod_a that requested 100% and used 0%, and pod_b that requested 0% and used 100%. According to the formulas you provide.

```
pod_a_delta_cost = max(0, 100) - 100 = 0
pod_b_delta_cost = max(100, 0) - 0 = 100

node_request_idle_cost = 0

pod_a_cost = 100 + 0 *... = 100
pod_b_cost = 0 + 0 * ... = 0.
```

pod_b in this case doesn't have any associated cost.

If the objective is to distribute 100% of the node cost among pods, irrespective of the node's actual usage, it can be accomplished through normalisation using the following formula:
```
pod_a_cost = node_cost * max(pod_a_used_cost, pod_a_requested_cost) / sum(max(pod_i_used_cost, pod_i_requested_cost))
```

But I think ensuring an accurate representation of node idle cost is crucial. For instance, in scenarios where pod1 uses 50% and requests 0%, and pod2 uses 0% and requests 50%, the current opencost model reports 100% utilisation of the node. However, in reality, 50% of both the requested and usage dimensions are available. I believe the logic for combining request and usage for the node should align with the logic used for pods: max(request, usage).

My suggestion is to make a slight adjustment to the formula and splitting the cost of only utilised portion of the node:
```
node_utilized_cost = max(node_used_cost, node_requested_cost)
pod_a_cost = node_utilized_cost * max(pod_a_used_cost, pod_a_requested_cost) / sum(max(pod_i_used_cost, pod_i_requested_cost))
```
In my opinion, providing the option for users to distribute idle resources among all pods proportionally to their cost or leave it untouched as a seperate record would be beneficial. This would allow for flexibility in generating various reports with different approaches.

**Comment by r2k1** [comment]
Here is prometheus recording rules file `recording_rules.yml` I use for such calculations (with some extras):

Note. It requires kube-state-metrics and won't work if opencost installed. Opencost metric names collides with KSM and it breaks some of the queries.

```yaml
groups:
  - name: resource_usage
    rules:
      - record: pod_resource_request
        # If a pod request metric doesn't have a node, such pod isn't allocated, and we are not interested in
        # It's a trick to perform union of metrics using "OR" operator,
        # without label replacement metric with matching set of labels will be ignored, with label_replace a unique set of labels is achieved
        # k8s logic for scheduling: https://github.com/kubernetes/kubernetes/blob/6d499ee9eacc4ccaef424dee5aef2f97ced215f0/pkg/scheduler/framework/plugins/noderesources/fit.go#L133-L175
        # total_request = max(sum(container_request), max(init_container_request))
        expr: >
          max by (node, namespace, pod, resource) (
              label_replace(
                sum by (pod, namespace, node, resource) (kube_pod_container_resource_requests{node!=""}),
                "union",
                "0",
                "",
                ""
              )
            or
              label_replace(kube_pod_init_container_resource_requests{node!=""}, "union", "1", "", "")
          )
      - record: pod_resource_used
        expr: >
          label_replace(
            (
                label_replace(
                  avg by (namespace, pod, instance) (rate(pod_cpu_usage_seconds_total[5m])),
                  "resource",
                  "cpu",
                  "",
                  ""
                )
              or
                label_replace(
                  avg by (namespace, pod, instance) (pod_memory_working_set_bytes),
                  "resource",
                  "memory",
                  "",
                  ""
                )
            ),
            "node",
            "$1",
            "instance",
            "(.*)"
          )
      - record: system_resource_reserved
        expr: >
          label_replace(
            (
              label_replace(
                (
                    (avg by (node, resource) (kube_node_status_capacity))
                  -
                    (avg by (node, resource) (kube_node_status_allocatable))
                ),
                "pod",
                "System",
                "",
                ""
              )
            ),
            "namespace",
            "System",
            "",
            ""
          )
      - record: pod_weight
        expr: >
          avg by (namespace, pod, node) (
            max by (namespace, pod, node, resource) (
                (
                      label_replace(pod_resource_used{resource=~"cpu|memory"}, "type", "used", "", "") > 0
                    or
                      label_replace(pod_resource_request{resource=~"cpu|memory"}, "type", "request", "", "") > 0
                  or
                    label_replace(system_resource_reserved{resource=~"cpu|memory"}, "type", "request", "", "") > 0
                )
              / on (node, resource) group_left ()
                (kube_node_status_capacity > 0)
            )
          )
      # We use a fake "System" pod, so we need to generate a fake owner for it to not lose the metric on join
      - record: pod_owner
        expr: >
          group by (pod, namespace, owner_kind, owner_name) (
            (
                (
                    (label_replace((kube_pod_owner{owner_kind="ReplicaSet"}), "replicaset", "$1", "owner_name", "(.+)"))
                  * on (replicaset, namespace) group_left (owner_kind, owner_name)
                    kube_replicaset_owner
                )
              or on (pod, namespace)
                (
                    (label_replace((kube_pod_owner{owner_kind="Job"}), "job_name", "$1", "owner_name", "(.+)"))
                  * on (job_name, namespace) group_left (owner_kind, owner_name)
                    kube_job_owner
                )
            )
            or on (pod, namespace)
                kube_pod_owner
            or
              label_replace(
                label_replace(
                  label_replace(
                    label_replace(
                      label_replace(vector(1), "owner_kind", "System", "", ""),
                      "owner_name",
                      "System",
                      "",
                      ""
                    ),
                    "namespace",
                    "System",
                    "",
                    ""
                  ),
                  "pod",
                  "System",
                  "",
                  ""
                ),
                "owner_kind",
                "System",
                "",
                ""
              )
          )
      - record: node_resource_used
        expr: >
          label_replace(
            (
                label_replace(rate(node_cpu_usage_seconds_total[5m]), "resource", "cpu", "", "")
              or
                label_replace(node_memory_working_set_bytes, "resource", "memory", "", "")
            ),
            "node",
            "$1",
            "instance",
            "(.*)"
          )
      - record: node_fraction
        expr: >
          max by (node) (
              label_replace(
                (sum by (node, resource) (pod_resource_request)) / on (resource, node) kube_node_status_allocatable,
                "type",
                "request",
                "",
                ""
              )
            or
              label_replace(
                node_resource_used / on (node, resource) kube_node_status_capacity,
                "type",
                "used",
                "",
                ""
              )
          )
      - record: pod_info
        # topk guarantees we have no more than a single metric per pod
        # duplicates may cause double counting or a query failure
        expr: >
          topk by (pod, namespace) (
            1,
            group without (uid, instance, job) (
                  (kube_pod_labels * on (pod, namespace) group_left (owner_kind, owner_name) pod_owner)
                or on (pod, namespace)
                  kube_pod_labels
              or on (pod, namespace)
                pod_owner
            )
          )
      # IMPORTANT METRIC
      - record: pod_node_fraction
        expr: >
          sum by (pod, namespace, node) (
                  (pod_weight / on (node) group_left () sum by (node) (pod_weight))
                * on (node) group_left ()
                  node_fraction
              )
            * on (pod, namespace) group_left (owner_kind, owner_name)
              pod_owner
          or
            label_replace(
              label_replace(
                label_replace(
                  label_replace(1 - node_fraction, "owner_kind", "Idle", "", ""),
                  "owner_name",
                  "Idle",
                  "",
                  ""
                ),
                "owner_kind",
                "Idle",
                "",
                ""
              ),
              "namespace",
              "Idle",
              "",
              ""
            )
```

**Comment by nikovacevic** [comment]
First of all, many thanks to @r2k1 for this thoughtful proposal. Let me begin my reply by quoting:

> [T]he proposal ensures that the total cost of all pods running on the node will match the cost of the node's utilization portion. This approach will never exceed 100% of the node cost and accurately represent the state of the node.

This seems to be the heart of the matter, and it's not clear that this is the primary goal of Opencost. This is certainly not what OpenCost does today, as @r2k1 adeptly demonstrates. But if we decide that this is the right primary goal, then this proposal seems to be an appropriate way to arrive at such an end.

However, there is another possible goal, which creates a competing concern:  determining the cost of a running container, based on the container's allocated resources and the price of those resources.

A question that aims to help disambiguate the two ends:

Consider two identical containers running simultaneously on two different nodes of the same type. The nodes' resource prices are identical. The containers' time running is identical. Should the two containers cost the same amount? If so, then there is a problem with this proposal, because it's possible the costs of these containers will become skewed by one node, and not the other, becoming over-provisioned in the way described above.

Today, OpenCost will give the same costs to these two containers, but one will run on a node with negative idle costs, which can then be shared optionally to adjust for the issue. Is that more fair? Opinions will differ. But it is at least valid to see it as fair from the perspective of two teams in an organization using Kubernetes, where one team carries a lower cost burden because the scheduler crowded one node more than another.

To summarize, OpenCost is trying to do multiple things that sometimes compete with each other. Two such things:
(1) Determine costs of containers, based on the resources allocated and the resource prices; and
(2) Determine accurate node costs, then distribute precisely 100% of the utilized node cost among the pods.

If (1) is more important than (2), then OpenCost works well today---and (2) can be achieved with idle cost, where idle cost will sometimes be negative. If (2) is more important than (1), then we should absolutely consider this proposal.

cc @dwbrown2 @AjayTripathy @teevans @kwombach

**Comment by r2k1** [comment]
My main challenge with the current model is splitting 100% of the cluster cost between containers **and idle** can be tricky. The term "idle" in opencost seems to have a different meaning.

**Comment by nikovacevic** [comment]
@r2k1 yes, there is an unfortunate naming ambiguity in the OpenCost spec. Kubecost engineers will usually understand "idle" to mean `cost of compute assets - cost of allocated compute` and is further partitioned by CPU, RAM, and GPU cost. E.g. if you have a node with (2.0 vCPU, 8.0 GiB RAM, 1.0 GPU), which ran for 1 hour, and the sum of resources allocated (`max(request, usage)`) to all workloads on that node totaled (1.5 vCPU-hours, 4.0 GiB-hours RAM, 0.0 GPU-hours) then the "idle" is (0.5, 4.0, 1.0), which can then be expressed in some unit of currency.

The spec also mentions "workload idle" which is most closely related to the Kubecost idea of "efficiency." I.e. how much of a given resource was allocated to a container, due to a resource request, but was not used. The same arithmetic and units apply, but rather than nodes and resources allocated to containers, it's resources requested by containers and resources used by containers -- (node:allocation :: request:usage).

> My main challenge with the current model is splitting 100% of the cluster cost between containers and idle can be tricky.

Which "idle" is your main concern here? What makes it tricky?

**Comment by r2k1** [comment]
My goal is to distribute node cost between all workloads running on the node. 

Using the query `sum(max(request, usage)) by (node)` on a busy cluster can yield results exceeding 100% of the node's resources. It can range from 0% to 200% depending on pod allocation. When using the formula `cost of compute assets - cost of allocated compute`, negative values are generated for idle instances. Interpreting and explaining negative values like -30% node is challenging.

For example:

```
Node 2.0 vCPU
Pod1 1.3 vCPU Request, 0.5 CPU Usage
Pod2 0.5 vCPU Request, 1.3 CPU Usage


pod1: 0.65 node cpu
pod2: 0.65 node cpu
node1 cost idle: -0.3 node cpu
node1 real idle: 0.2 node cpu

```

**Comment by nikovacevic** [comment]
I see, so you're referring to the same "idle" -- not "workload idle." Yeah, I don't think I have any more to contribute to this conversation beyond what I [stated previously](https://github.com/opencost/opencost/issues/1815#issuecomment-1568971454).

We'll just have to wait for one of those tagged to reply from a founder/product perspective.

**Comment by AjayTripathy** [comment]
Hey just wanted to weigh in here with some historical context on why we accepted the possibility of negative idle or overallocating costs-- it's to compare workloads across architectures. It has to do with @nikovacevic point here:
> Consider two identical containers running simultaneously on two different nodes of the same type. The nodes' resource prices are identical. The containers' time running is identical. Should the two containers cost the same amount? If so, then there is a problem with this proposal, because it's possible the costs of these containers will become skewed by one node, and not the other, becoming over-provisioned in the way described above.

Now, imagine you want to compare these containers with, say, a job that does something similar on...elastic mapreduce. The common denominator, even across multiple architectures, is cost. If the reported cost is influenced by other containers on a node, then cost becomes a function of some specific things in the kubernetes scheduler and not just a function of cpu/gpu/ram.

That being said, we have the underlying data to compute both sides of the coin, so I'm not opposed to the ability to compute both models--or if I'm missing something please do let me know.

**Comment by 123BLiN** [comment]
@r2k1 if it is not a secret - what are you doing then with this data you got from prometheus recording rules?

We are doing something similar with node Idle costs in the post-processing python script - so we query kubecost API then distribute node Idle costs to workloads using the same formula i.e. `max(request, usage)` 

given your recording rules it seems we might get this `pod_node_fraction` directly from prometheus (which is a bit of a news to me - never thought such complicated calculations can be expressed with promql tbh) 

then theoretically we can get the CUR report from AWS and process two datasets to get the real pod price, and the more all pods running in sum should give us accurate pricing that equals to the AWS costs explorer numbers including all savings plans spots etc. 

 ```
To summarize, OpenCost is trying to do multiple things that sometimes compete with each other. Two such things:
(1) Determine costs of containers, based on the resources allocated and the resource prices; and
(2) Determine accurate node costs, then distribute precisely 100% of the utilized node cost among the pods.
```
in our case 2) is definitely more important - we do understand the value of Idle and additional info it gives but this info we can get from anywhere, accurate pod/node and total cluster pricing is crucial

**Comment by r2k1** [comment]
@123BLiN 

Currently, I'm not using this anymore, but the original goal was the same: to join the Prometheus data with cloud cost to get the cost of each pod.

You don't have to use recording rules. It's possible to build a huge and complex query with a lot of duplication and some complex tricks (though I'm not sure about efficiency on large clusters). VictoriaMetrics has a much nicer query language with support for WITH statements that significantly simplify complex queries.

You can mimic OpenCost behavior in Prometheus. A distinctive OpenCost feature is the `1 CPU core ≈ 7GiB of memory` formula that is used to combine memory and CPU.

**Comment by r2k1** [comment]
Something like this:

```
WITH (
  pod_avg_used_memory = avg_over_time(pod_memory_working_set_bytes)[1d],
  pod_avg_requested_memory = avg_over_time(pod_request_memory)[1d],
  pod_avg_used_cpu = rate(pod_cpu_usage_seconds_total[1d]),
  pod_avg_requested_cpu = avg_over_time(pod_request_cpu[1d]),
  pod_avg_requested_gpu = avg_over_time(pod_request_gpu[1d]),
  pod_memory = (
    max(
      label_set(pod_avg_used_memory, "t", "used")
        or
      label_set(pod_avg_requested_memory, "t", "request")
    ) by(pod,namespace,node)
  ),
  pod_cpu = (
    max(
      label_set(pod_avg_used_cpu, "t", "used")
        or
      label_set(pod_avg_requested_cpu, "t", "request")
    ) by(pod,namespace,node)
  ),
  pod_compute = (
    sum(
      (
        label_set(pod_memory / 7Gi, "t", "memory")
          or
        label_set(pod_cpu, "t", "cpu")
      )
        or
      label_set(pod_avg_requested_gpu * 30, "t", "gpu")
    ) by(namespace,pod,node)
  ),
  pod_hours = duration_over_time(pod[1d], 5m) / 3600,
  controller_compute_hours = (
    sum(
      pod_compute
        * on(namespace,pod,node) group_left(controller_kind,controller_name)
      pod_hours
    ) by(namespace,node,controller_name,controller_kind)
  ),
  node_capacity = (
    sum(
      (
        label_set(avg_over_time(node_capacity_memory[1d]) / 7Gi, "t", "memory")
          or
        label_set(avg_over_time(node_capacity_cpu[1d]), "t", "cpu")
      )
        or
      label_set(avg_over_time(node_capacity_gpu[1d]) * 30, "t", "gpu")
    ) by(node)
  ),
  controller_node_hours  = (
    controller_compute_hours / on(node) group_left() node_capacity
  ),
)
controller_node_hours
```

**Comment by r2k1** [comment]
I don't believe it will go anywhere, so I'm closing the issue
